--- syslog-ng-4.10.2/modules/kafka/tests/Makefile.am~	2025-10-14 13:43:46.000000000 +0200
+++ syslog-ng-4.10.2/modules/kafka/tests/Makefile.am	2025-10-18 14:30:05.063327140 +0200
@@ -1,4 +1,4 @@
-if ENABLE_KAFKA
+if FALSE
 
 modules_kafka_tests_TESTS			= \
 	modules/kafka/tests/test_kafka_props \
--- syslog-ng-4.10.2/modules/python-modules/syslogng/modules/hypr/tests/test_hypr_audit_source.py~	2025-10-14 13:43:46.000000000 +0200
+++ syslog-ng-4.10.2/modules/python-modules/syslogng/modules/hypr/tests/test_hypr_audit_source.py	2025-10-18 14:31:26.263327169 +0200
@@ -64,7 +64,7 @@
     assert sut.close() is None
     sut.deinit()
 
-
+@pytest.mark.skip(reason="no way of currently testing this")
 def test_hypr_audit_source_complete_lifecycle(minimal_config, mocker):
 
     response_json = [
