--- syslog-ng-4.8.0/configure.ac~	2024-08-01 19:27:51.000000000 +0200
+++ syslog-ng-4.8.0/configure.ac	2024-08-01 19:28:54.109970068 +0200
@@ -1036,16 +1036,18 @@ dnl GLib headers/libraries
 dnl ***************************************************************************
 
 GLIB_ADDONS="gmodule-2.0 gthread-2.0"
-PKG_CHECK_MODULES(GLIB, glib-2.0 >= $GLIB_MIN_VERSION $GLIB_ADDONS,,)
 
 if test "$linking_mode" != "dynamic"; then
+        PKG_CHECK_MODULES_STATIC(GLIB, glib-2.0 >= $GLIB_MIN_VERSION $GLIB_ADDONS,,)
 	# strip out -ldl & -lrt as it cannot be linked statically
-	GLIB_LIBS=`echo $GLIB_LIBS | tr ' ' '\n' | egrep -v "^(-ldld?)|(-lrt)$" | tr '\n' ' '`
+        GLIB_LIBS=`echo $GLIB_LIBS | tr ' ' '\n' | egrep -v "^(-ldld?)|(-lrt)|(-lm)|(-lpcre2-8)$" | tr '\n' ' '`
 
 	old_LIBS=$LIBS
 	LIBS="$LD_START_STATIC $GLIB_LIBS $LD_END_STATIC $BASE_LIBS"
 	AC_CHECK_FUNC(g_hash_table_new, blb_cv_static_glib=yes, blb_cv_static_glib=no)
 	LIBS=$old_LIBS
+else
+        PKG_CHECK_MODULES(GLIB, glib-2.0 >= $GLIB_MIN_VERSION $GLIB_ADDONS,,)
 fi
 
 old_CPPFLAGS=$CPPFLAGS
