--- syslog-ng-3.34.1/modules/cryptofuncs/cryptofuncs.c~	2021-09-10 13:01:49.000000000 +0200
+++ syslog-ng-3.34.1/modules/cryptofuncs/cryptofuncs.c	2021-10-11 22:51:20.705912103 +0200
@@ -162,7 +162,6 @@
   TEMPLATE_FUNCTION_PLUGIN(tf_hash, "sha1"),
   TEMPLATE_FUNCTION_PLUGIN(tf_hash, "sha256"),
   TEMPLATE_FUNCTION_PLUGIN(tf_hash, "sha512"),
-  TEMPLATE_FUNCTION_PLUGIN(tf_hash, "md4"),
   TEMPLATE_FUNCTION_PLUGIN(tf_hash, "md5"),
 };
 
--- syslog-ng-3.34.1/modules/cryptofuncs/tests/test_cryptofuncs.c~	2021-09-10 13:01:49.000000000 +0200
+++ syslog-ng-3.34.1/modules/cryptofuncs/tests/test_cryptofuncs.c	2021-10-11 22:51:41.905839413 +0200
@@ -50,7 +50,6 @@
   assert_template_format("$(sha1 bar)", "62cdb7020ff920e5aa642c3d4066950dd1f01f4d");
   assert_template_format("$(md5 foo)", "acbd18db4cc2f85cedef654fccc4a4d8");
   assert_template_format("$(hash foo)", "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae");
-  assert_template_format("$(md4 foo)", "0ac6700c491d70fb8650940b1ca1e4b2");
   assert_template_format("$(sha256 foo)", "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae");
   assert_template_format("$(sha512 foo)",
                          "f7fbba6e0636f890e56fbbf3283e524c6fa3204ae298382d624741d0dc6638326e282c41be5e4254d8820772c5518a2c5a8c0c7f7eda19594a7eb539453e1ed7");
--- syslog-ng-3.34.1/modules/python/pylib/syslogng/debuggercli/syslognginternals.py~	2021-09-10 13:01:49.000000000 +0200
+++ syslog-ng-3.34.1/modules/python/pylib/syslogng/debuggercli/syslognginternals.py	2021-10-11 22:52:29.489009746 +0200
@@ -52,7 +52,6 @@
         "sha1",
         "sha256",
         "sha512",
-        "md4",
         "md5",
         "format-json",
         "grep",
