diff -Nur syslog-ng-1.6.0rc3.orig/configure.in syslog-ng-1.6.0rc3.chng/configure.in
--- syslog-ng-1.6.0rc3.orig/configure.in	Fri Apr 25 08:58:08 2003
+++ syslog-ng-1.6.0rc3.chng/configure.in	Mon Jul 28 15:22:26 2003
@@ -150,9 +150,9 @@
 AC_CHECK_LIB(socket, socket)
 AC_CHECK_LIB(nsl, gethostbyname)
 AC_CHECK_FUNCS(select snprintf vsnprintf strerror inet_aton strncpy getutent)
-AC_CHECK_FUNC(getopt_long,,[LIBOBJS="getopt.o getopt1.o $LIBOBJS"])
-AC_CHECK_FUNC(strcasecmp,,[LIBOBJS="strcasecmp.o $LIBOBJS"])
-AC_CHECK_FUNC(strptime,,[LIBOBJS="strptime.o $LIBOBJS"])
+AC_CHECK_FUNC(getopt_long,,[AC_LIBOBJ([getopt]); AC_LIBOBJ([getopt1])])
+AC_CHECK_FUNC(strcasecmp,,[AC_LIBOBJ([strcasecmp])])
+AC_CHECK_FUNC(strptime,,[AC_LIBOBJ([strptime])])
 
 old_LIBS=$LIBS
 
@@ -186,7 +186,7 @@
 
 if test "x$ac_cv_func_snprintf" = "xno" -o \
 	"x$ac_cv_func_vsnprintf" = "xno"; then
-	LIBOBJS="snprintf.o $LIBOBJS"
+	AC_LIBOBJ([snprintf])
 fi
 
 if test "x$ac_cv_lib_door_door_create" = "xyes"; then
@@ -278,7 +278,6 @@
 
 LIBS="$LIBS $LEXLIB $LIBOL_LIBS"
 AC_SUBST(YFLAGS)
-AC_SUBST(LIBOBJS)
 
 AC_SUBST(MAKE_CLASS)
 AC_DEFINE_UNQUOTED(PATH_SYSCONFDIR, "`eval echo $sysconfdir`")
