--- syslog-ng-3.6.2/contrib/systemd/syslog-ng.service~	2014-12-15 16:48:45.000000000 +0200
+++ syslog-ng-3.6.2/contrib/systemd/syslog-ng.service	2015-06-01 08:06:32.632888085 +0300
@@ -4,7 +4,9 @@ Documentation=man:syslog-ng(8)
 
 [Service]
 Type=notify
-ExecStart=/usr/sbin/syslog-ng -F $SYSLOGNG_OPTS
+EnvironmentFile=-/etc/sysconfig/syslog-ng
+ExecStartPre=-/bin/systemctl stop systemd-kmsg-syslogd.service
+ExecStart=/sbin/syslog-ng -F -f /etc/syslog-ng/syslog-ng.conf -p /var/run/syslogd.pid --worker-threads=1024 $SYSLOGNG_OPTS
 ExecReload=/bin/kill -HUP $MAINPID
 EnvironmentFile=-/etc/default/syslog-ng
 EnvironmentFile=-/etc/sysconfig/syslog-ng
