--- syslog-ng-3.3.6/configure.in.orig	2013-02-03 16:17:57.687433055 +0100
+++ syslog-ng-3.3.6/configure.in	2013-02-03 16:30:24.357417494 +0100
@@ -67,7 +67,7 @@
 pidfiledir='${localstatedir}'
 moduledir='${exec_prefix}/lib/syslog-ng'
 
-AM_CONFIG_HEADER(config.h)
+AC_CONFIG_HEADERS(config.h)
 
 dnl ***************************************************************************
 dnl Arguments
@@ -211,7 +211,6 @@
 dnl ***************************************************************************
 dnl Checks for programs.
 AC_PROG_CC
-AM_PROG_CC_STDC
 AC_PROG_CC_C99
 if test "x$ac_cv_prog_cc_c99" = "xno"; then
 	AC_MSG_ERROR("C99 standard compliant C compiler required. Try GCC 3.x or later.")
