--- syslog-ng-1.9.9/src/cfg-grammar.y.orig	2006-02-26 09:31:44.000000000 +0100
+++ syslog-ng-1.9.9/src/cfg-grammar.y	2006-03-25 23:05:27.000126000 +0100
@@ -787,15 +787,11 @@
 	      msg_error("Warning: Unknown priority level",
                         evt_tag_str("priority", $1),
                         NULL);
-	    else
-	      r1 = sl_levels[r1].value;
 	    r2 = syslog_lookup_level($3);
 	    if (r2 == -1)
 	      msg_error("Warning: Unknown priority level",
                         evt_tag_str("priority", $1),
                         NULL);
-	    else
-	      r2 = sl_levels[r2].value;
 	    if (r1 != -1 && r2 != -1)
 	      $$ = syslog_make_range(r1, r2); 
 	    else
@@ -814,7 +810,7 @@
 	        $$ = 0;
 	      }
 	    else
-	      $$ = 1 << sl_levels[n].value; 
+	      $$ = (1 << n);
 	    free($1); 
 	  }
 	;
