
 -FIXME: drop this patch and just fix PATH_LOCALSTATEDIR ?

diff -urNp -x '*.orig' syslog-ng-3.29.1.org/lib/syslog-ng.h syslog-ng-3.29.1/lib/syslog-ng.h
--- syslog-ng-3.29.1.org/lib/syslog-ng.h	2020-08-28 14:08:43.000000000 +0200
+++ syslog-ng-3.29.1/lib/syslog-ng.h	2021-02-28 14:25:45.374311232 +0100
@@ -42,7 +42,7 @@
 #if SYSLOG_NG_ENABLE_ENV_WRAPPER
 #define PATH_SYSLOGNG           SYSLOG_NG_PATH_LIBEXECDIR "/syslog-ng"
 #endif
-#define PATH_PERSIST_CONFIG     SYSLOG_NG_PATH_LOCALSTATEDIR "/syslog-ng.persist"
+#define PATH_PERSIST_CONFIG     "/var/lib/syslog-ng/syslog-ng.persist"
 
 typedef struct _LogPipe LogPipe;
 typedef struct _LogMessage LogMessage;
diff -urNp -x '*.orig' syslog-ng-3.29.1.org/scripts/update-patterndb.in syslog-ng-3.29.1/scripts/update-patterndb.in
--- syslog-ng-3.29.1.org/scripts/update-patterndb.in	2020-08-10 20:04:39.000000000 +0200
+++ syslog-ng-3.29.1/scripts/update-patterndb.in	2021-02-28 14:25:45.374311232 +0100
@@ -3,6 +3,6 @@ prefix=@prefix@
 exec_prefix=@exec_prefix@
 bindir=@bindir@
 sysconfdir=@sysconfdir@
-localstatedir=@localstatedir@
+localstatedir=/var/lib/syslog-ng
 
 ${bindir}/pdbtool merge -r --glob \*.pdb -D ${sysconfdir}/patterndb.d -p ${localstatedir}/patterndb.xml
