--- syslog-ng-3.0.2/tests/unit/test_zone.c~	2009-05-08 10:52:59.000000000 +0200
+++ syslog-ng-3.0.2/tests/unit/test_zone.c	2009-05-12 16:30:34.000000000 +0200
@@ -468,19 +468,6 @@
   TEST_ASSERT(test_timezone(now, "MST7MDT"));
   TEST_ASSERT(test_timezone(now, "GB-Eire"));
   TEST_ASSERT(test_timezone(now, "PRC"));
-  TEST_ASSERT(test_timezone(now, "SystemV/CST6CDT"));
-  TEST_ASSERT(test_timezone(now, "SystemV/CST6"));
-  TEST_ASSERT(test_timezone(now, "SystemV/PST8"));
-  TEST_ASSERT(test_timezone(now, "SystemV/YST9"));
-  TEST_ASSERT(test_timezone(now, "SystemV/EST5"));
-  TEST_ASSERT(test_timezone(now, "SystemV/AST4"));
-  TEST_ASSERT(test_timezone(now, "SystemV/AST4ADT"));
-  TEST_ASSERT(test_timezone(now, "SystemV/YST9YDT"));
-  TEST_ASSERT(test_timezone(now, "SystemV/EST5EDT"));
-  TEST_ASSERT(test_timezone(now, "SystemV/PST8PDT"));
-  TEST_ASSERT(test_timezone(now, "SystemV/MST7MDT"));
-  TEST_ASSERT(test_timezone(now, "SystemV/HST10"));
-  TEST_ASSERT(test_timezone(now, "SystemV/MST7"));
   TEST_ASSERT(test_timezone(now, "Arctic/Longyearbyen"));
   TEST_ASSERT(test_timezone(now, "Cuba"));
   TEST_ASSERT(test_timezone(now, "Israel"));
