--- syslog-ng-3.5.2/contrib/systemd/syslog-ng.service~	2013-09-17 17:43:11.000000000 +0200
+++ syslog-ng-3.5.2/contrib/systemd/syslog-ng.service	2013-12-01 22:50:06.249492652 +0100
@@ -3,9 +3,11 @@
 Documentation=man:syslog-ng(8)
 
 [Service]
+EnvironmentFile=-/etc/sysconfig/syslog-ng
 Type=notify
 Sockets=syslog.socket
-ExecStart=/usr/sbin/syslog-ng -F
+ExecStartPre=-/bin/systemctl stop systemd-kmsg-syslogd.service
+ExecStart=/sbin/syslog-ng -F -f /etc/syslog-ng/syslog-ng.conf --worker-threads=1024
 ExecReload=/bin/kill -HUP $MAINPID
 StandardOutput=null
 Restart=on-failure
