--- syslog-ng-2.0.3/configure.in.orig	2007-03-04 10:58:00.000000000 +0100
+++ syslog-ng-2.0.3/configure.in	2007-04-03 23:53:53.219462987 +0200
@@ -265,9 +265,8 @@
 AC_CHECK_LIB(door, door_create, BASE_LIBS="$BASE_LIBS -ldoor")
 AC_CHECK_LIB(socket, socket, BASE_LIBS="$BASE_LIBS -lsocket")
 AC_CHECK_LIB(rt, nanosleep, BASE_LIBS="$BASE_LIBS -lrt")
-AC_CHECK_LIB(nsl, gethostbyname, BASE_LIBS="$BASE_LIBS -lnsl")
+AC_CHECK_LIB(resolv, gethostbyname, BASE_LIBS="$BASE_LIBS -lresolv")
 AC_CHECK_LIB(regex, regexec, REGEX_LIBS="-lregex")
-AC_CHECK_LIB(resolv, res_init, RESOLV_LIBS="-lresolv")
 
 if test "x$enable_linux_caps" = "xyes" -o "x$enable_linux_caps" = "xauto"; then
 	AC_CHECK_LIB(cap, cap_set_proc, LIBCAP_LIBS="-lcap")
@@ -674,7 +674,7 @@
 	CORE_DEPS_LIBS="$DEPS_LIBS"
 	SYSLOGNG_LINK='$(LINK)'
 else
-	DEPS_LIBS="$LIBS $BASE_LIBS $RESOLV_LIBS $LD_START_STATIC -Wl,--whole-archive $GLIB_LIBS $EVTLOG_LIBS $PCRE_LIBS $REGEX_LIBS -Wl,--no-whole-archive $LD_END_STATIC $LIBCAP_LIBS $DL_LIBS "
+	DEPS_LIBS="$LIBS $BASE_LIBS $RESOLV_LIBS $LD_START_STATIC -Wl,--whole-archive $GLIB_LIBS $EVTLOG_LIBS $PCRE_LIBS $REGEX_LIBS -Wl,--no-whole-archive $LD_END_STATIC $LIBCAP_LIBS $DL_LIBS -lrt "
 	CORE_DEPS_LIBS=""
 	# bypass libtool in case we want to do mixel linking
 	SYSLOGNG_LINK='$(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@'
--- syslog-ng-3.2.2/modules/dbparser/Makefile.am~	2011-01-15 17:14:39.000000000 +0100
+++ syslog-ng-3.2.2/modules/dbparser/Makefile.am	2011-01-17 20:06:00.103986725 +0100
@@ -17,7 +17,7 @@
 	timerwheel.c timerwheel.h \
 	patternize.c patternize.h
 libsyslog_ng_patterndb_la_LDFLAGS = -no-undefined
-libsyslog_ng_patterndb_la_LIBADD =  @OPENSSL_LIBS@ @DEPS_LIBS@
+libsyslog_ng_patterndb_la_LIBADD =  $(top_builddir)/lib/libsyslog-ng.la @OPENSSL_LIBS@ @DEPS_LIBS@
 
 module_LTLIBRARIES = libdbparser.la
 libdbparser_la_SOURCES = \
--- syslog-ng-3.2.2/configure.in~	2011-02-23 08:46:17.152179679 +0100
+++ syslog-ng-3.2.2/configure.in	2011-02-23 08:53:26.506748469 +0100
@@ -458,7 +458,7 @@
 	GLIB_LIBS=`echo $GLIB_LIBS | tr ' ' '\n' | egrep -v "^(-ldld?)|(-lrt)$" | tr '\n' ' '`
 
 	old_LIBS=$LIBS
-	LIBS="$LD_START_STATIC $GLIB_LIBS $LD_END_STATIC"
+	LIBS="$LD_START_STATIC $GLIB_LIBS $LD_END_STATIC -ldl -lrt"
 	AC_CHECK_FUNC(g_hash_table_new, blb_cv_static_glib=yes, blb_cv_static_glib=no)
 	LIBS=$old_LIBS
 fi
--- syslog-ng-3.2.2/configure.in~	2011-02-23 08:53:58.767670120 +0100
+++ syslog-ng-3.2.2/configure.in	2011-02-23 09:02:25.059826526 +0100
@@ -675,7 +675,7 @@
 	SYSLOGNG_LINK='$(LINK)'
 else
 	DEPS_LIBS="$LIBS $BASE_LIBS $RESOLV_LIBS $LD_START_STATIC -Wl,--whole-archive $GLIB_LIBS $EVTLOG_LIBS $PCRE_LIBS $REGEX_LIBS -Wl,--no-whole-archive $LD_END_STATIC $LIBCAP_LIBS $DL_LIBS -lrt "
-	CORE_DEPS_LIBS=""
+	CORE_DEPS_LIBS="$LD_START_STATIC -Wl,--whole-archive $GLIB_LIBS $EVTLOG_LIBS -Wl,--no-whole-archive $LD_END_STATIC $LIBCAP_LIBS $DL_LIBS -lrt"
 	# bypass libtool in case we want to do mixel linking
 	SYSLOGNG_LINK='$(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@'
 fi
