--- syslog-ng-3.2.4/configure.in~	2011-05-04 17:45:07.000000000 +0200
+++ syslog-ng-3.2.4/configure.in	2011-05-30 20:09:43.852080011 +0200
@@ -770,7 +770,7 @@
 else
 	SYSLOGNG_DEPS_LIBS="$LIBS $BASE_LIBS $RESOLV_LIBS $LD_START_STATIC -Wl,--whole-archive $GLIB_LIBS $EVTLOG_LIBS $PCRE_LIBS $REGEX_LIBS -Wl,--no-whole-archive $LD_END_STATIC $LIBCAP_LIBS $DL_LIBS "
         TOOL_DEPS_LIBS="$LIBS $BASE_LIBS $GLIB_LIBS $EVTLOG_LIBS $RESOLV_LIBS $LIBCAP_LIBS $PCRE_LIBS $REGEX_LIBS $DL_LIBS"
-	CORE_DEPS_LIBS=""
+	CORE_DEPS_LIBS="$LD_START_STATIC -Wl,--whole-archive $GLIB_LIBS $EVTLOG_LIBS -Wl,--no-whole-archive $LD_END_STATIC $LIBCAP_LIBS"
 
 	# bypass libtool in case we want to do mixed linking because it
 	# doesn't support -Wl,-Bstatic -Wl,-Bdynamic on a per-library basis.
