--- syslog-ng-4.8.0/configure.ac~	2024-07-23 13:42:10.000000000 +0200
+++ syslog-ng-4.8.0/configure.ac	2024-08-01 19:20:42.196636751 +0200
@@ -900,11 +900,10 @@ dnl ************************************
 dnl Checks for libraries
 AC_CHECK_LIB(m, round, BASE_LIBS="$BASE_LIBS -lm")
 AC_CHECK_LIB(door, door_create, BASE_LIBS="$BASE_LIBS -ldoor")
-AC_CHECK_LIB(socket, socket, BASE_LIBS="$BASE_LIBS -lsocket")
+AC_CHECK_FUNC([socket], [:], [AC_CHECK_LIB(socket, socket, BASE_LIBS="$BASE_LIBS -lsocket")])
 AC_CHECK_LIB(rt, nanosleep, BASE_LIBS="$BASE_LIBS -lrt")
-AC_CHECK_FUNC(gethostbyname, [], AC_CHECK_LIB(nsl, gethostbyname, BASE_LIBS="$BASE_LIBS -lnsl"))
-AC_CHECK_LIB(regex, regexec, REGEX_LIBS="-lregex")
-AC_CHECK_LIB(resolv, res_init, RESOLV_LIBS="-lresolv")
+AC_CHECK_FUNC([gethostbyname], [:], [AC_CHECK_LIB(nsl, gethostbyname, BASE_LIBS="$BASE_LIBS -lnsl")])
+AC_CHECK_FUNC([regexec], [:], [AC_CHECK_LIB(regex, regexec, REGEX_LIBS="-lregex")])
 
 CHECK_FUNCS_AND_DEFINE([
     fmemopen
