diff -dur -x '*~' syslog-ng-syslog-ng-3.14.1.orig/contrib/systemd/syslog-ng@.service syslog-ng-syslog-ng-3.14.1/contrib/systemd/syslog-ng@.service
--- syslog-ng-syslog-ng-3.14.1.orig/contrib/systemd/syslog-ng@.service	2018-02-26 15:32:28.000000000 +0100
+++ syslog-ng-syslog-ng-3.14.1/contrib/systemd/syslog-ng@.service	2018-03-30 15:26:03.000000000 +0200
@@ -7,11 +7,14 @@
 Type=notify
 EnvironmentFile=-/etc/default/syslog-ng@%i
 EnvironmentFile=-/etc/sysconfig/syslog-ng@%i
-ExecStart=/usr/sbin/syslog-ng -F $OTHER_OPTIONS --cfgfile $CONFIG_FILE --control $CONTROL_FILE --persist-file $PERSIST_FILE --pidfile $PID_FILE
+EnvironmentFile=-/etc/sysconfig/syslog-ng
+ExecStartPre=-/bin/systemctl stop systemd-kmsg-syslogd.service
+ExecStart=/usr/sbin/syslog-ng -F $OTHER_OPTIONS --cfgfile /etc/syslog-ng/syslog-ng.conf -p /var/run/syslogd.pid --worker-threads=1024 $SYSLOGNG_OPTS
 ExecReload=/bin/kill -HUP $MAINPID
 StandardOutput=journal
 StandardError=journal
 Restart=on-failure
 
 [Install]
+Alias=syslog-ng.service
 WantedBy=multi-user.target
