FIXME: drop this patch and just fix PATH_LOCALSTATEDIR ?


--- syslog-ng-3.6.2/lib/syslog-ng.h~	2014-12-15 16:48:45.000000000 +0200
+++ syslog-ng-3.6.2/lib/syslog-ng.h	2015-06-01 08:04:18.542588343 +0300
@@ -42,7 +42,7 @@
 #if SYSLOG_NG_ENABLE_ENV_WRAPPER
 #define PATH_SYSLOGNG           SYSLOG_NG_PATH_LIBEXECDIR "/syslog-ng"
 #endif
-#define PATH_PERSIST_CONFIG     SYSLOG_NG_PATH_LOCALSTATEDIR "/syslog-ng.persist"
+#define PATH_PERSIST_CONFIG     "/var/lib/syslog-ng/syslog-ng.persist"
 
 typedef struct _LogPipe LogPipe;
 typedef struct _LogMessage LogMessage;
--- syslog-ng-3.2.2/scripts/update-patterndb.in~	2010-11-20 09:47:33.000000000 +0100
+++ syslog-ng-3.2.2/scripts/update-patterndb.in	2011-01-27 20:43:14.656000204 +0100
@@ -3,6 +3,6 @@
 exec_prefix=@exec_prefix@
 bindir=@bindir@
 sysconfdir=@sysconfdir@
-localstatedir=@localstatedir@
+localstatedir=/var/lib/syslog-ng
 
 ${bindir}/pdbtool merge -r --glob *.pdb -D ${sysconfdir}/patterndb.d -p ${localstatedir}/patterndb.xml
