--- syslog-ng-3.3.6/contrib/systemd/syslog-ng.service~	2012-08-15 10:06:08.000000000 +0200
+++ syslog-ng-3.3.6/contrib/systemd/syslog-ng.service	2013-02-03 14:06:49.380549410 +0100
@@ -3,8 +3,10 @@
 Documentation=man:syslog-ng(8)
 
 [Service]
+EnvironmentFile=-/etc/sysconfig/syslog-ng
 Sockets=syslog.socket
-ExecStart=/usr/sbin/syslog-ng -F
+ExecStartPre=-/bin/systemctl stop systemd-kmsg-syslogd.service
+ExecStart=/sbin/syslog-ng -F -f /etc/syslog-ng/syslog-ng.conf --worker-threads=1024
 ExecReload=/bin/kill -HUP $MAINPID
 StandardOutput=null
 

