--- syslog-ng-3.3.1/configure.in.orig	2011-09-29 08:35:39.000000000 +0200
+++ syslog-ng-3.3.1/configure.in	2011-11-11 10:37:51.371629795 +0100
@@ -465,11 +465,10 @@
 dnl ***************************************************************************
 dnl Checks for libraries
 AC_CHECK_LIB(door, door_create, BASE_LIBS="$BASE_LIBS -ldoor")
-AC_CHECK_LIB(socket, socket, BASE_LIBS="$BASE_LIBS -lsocket")
+AC_CHECK_FUNC([socket], [:], [AC_CHECK_LIB(socket, socket, BASE_LIBS="$BASE_LIBS -lsocket")])
 AC_CHECK_LIB(rt, nanosleep, BASE_LIBS="$BASE_LIBS -lrt")
-AC_CHECK_LIB(nsl, gethostbyname, BASE_LIBS="$BASE_LIBS -lnsl")
-AC_CHECK_LIB(regex, regexec, REGEX_LIBS="-lregex")
-AC_CHECK_LIB(resolv, res_init, RESOLV_LIBS="-lresolv")
+AC_CHECK_FUNC([gethostbyname], [:], [AC_CHECK_LIB(nsl, gethostbyname, BASE_LIBS="$BASE_LIBS -lnsl")])
+AC_CHECK_FUNC([regexec], [:], [AC_CHECK_LIB(regex, regexec, REGEX_LIBS="-lregex")])
 
 
 if test "x$enable_linux_caps" = "xyes" -o "x$enable_linux_caps" = "xauto"; then
