--- syslog-ng-3.2.4/lib/Makefile.am~	2011-02-05 16:01:43.000000000 +0100
+++ syslog-ng-3.2.4/lib/Makefile.am	2011-05-30 19:22:15.813931407 +0200
@@ -3,7 +3,7 @@
 export top_srcdir
 
 lib_LTLIBRARIES = libsyslog-ng.la
-libsyslog_ng_la_LIBADD = @CORE_DEPS_LIBS@
+libsyslog_ng_la_LIBADD = @CORE_DEPS_LIBS@ @GLIB_LIBS@ @EVTLOG_LIBS@ -lcap
 libsyslog_ng_la_LDFLAGS = -no-undefined
 
 libsyslog_ng_la_SOURCES = \

