--- syslog-ng-4.8.0/contrib/systemd/syslog-ng@.service.orig	2024-08-01 19:21:06.909970083 +0200
+++ syslog-ng-4.8.0/contrib/systemd/syslog-ng@.service	2024-08-01 19:23:37.569970075 +0200
@@ -7,9 +7,13 @@ After=network.target network-online.targ
 
 [Service]
 Type=notify
+Environment=CONFIG_FILE=/etc/syslog-ng/syslog-ng.conf
+Environment=PID_FILE=/var/run/syslogd.pid
+Environment=PERSIST_FILE=/var/lib/syslog-ng/syslog-ng.persist
+Environment=CONTROL_FILE=/var/lib/syslog-ng/syslog-ng.ctl
 EnvironmentFile=-/etc/default/syslog-ng@%i
 EnvironmentFile=-/etc/sysconfig/syslog-ng@%i
-ExecStart=/usr/sbin/syslog-ng -F $OTHER_OPTIONS --cfgfile $CONFIG_FILE --control $CONTROL_FILE --persist-file $PERSIST_FILE --pidfile $PID_FILE
+ExecStart=/sbin/syslog-ng -F $OTHER_OPTIONS --cfgfile $CONFIG_FILE --control $CONTROL_FILE --persist-file $PERSIST_FILE --pidfile $PID_FILE $SYSLOGNG_OPTS
 ExecReload=/bin/kill -HUP $MAINPID
 StandardOutput=journal
 StandardError=journal
