--- syslog-ng-3.2.2/lib/syslog-ng.h~	2011-01-04 12:57:06.000000000 +0100
+++ syslog-ng-3.2.2/lib/syslog-ng.h	2011-01-17 19:31:46.933293237 +0100
@@ -47,10 +47,10 @@
 #define PATH_SYSLOGNG           PATH_LIBEXECDIR "/syslog-ng"
 #endif
 
-#define PATH_PERSIST_CONFIG     PATH_LOCALSTATEDIR "/syslog-ng.persist"
-#define PATH_QDISK              PATH_LOCALSTATEDIR
-#define PATH_PATTERNDB_FILE     PATH_LOCALSTATEDIR "/patterndb.xml"
-#define PATH_XSDDIR             PATH_DATADIR "/xsd"
+#define PATH_PERSIST_CONFIG     "/var/lib/syslog-ng/syslog-ng.persist"
+#define PATH_QDISK              "/var/lib/syslog-ng"
+#define PATH_PATTERNDB_FILE     "/var/lib/syslog-ng/patterndb.xml"
+#define PATH_XSDDIR             "/var/lib/syslog-ng/xsd"
 
 #define LOG_PRIORITY_LISTEN 0
 #define LOG_PRIORITY_READER 0
