diff -urNp -x '*.orig' syslog-ng-3.29.1.org/configure.ac syslog-ng-3.29.1/configure.ac
--- syslog-ng-3.29.1.org/configure.ac	2021-02-28 14:27:09.946954361 +0100
+++ syslog-ng-3.29.1/configure.ac	2021-02-28 14:27:10.230296549 +0100
@@ -1775,9 +1775,9 @@ if test "x$linking_mode" = "xdynamic"; t
 	# syslog-ng binary is linked with the default link command (e.g. libtool)
 	SYSLOGNG_LINK='$(LINK)'
 else
-	SYSLOGNG_DEPS_LIBS="$LIBS $BASE_LIBS $RESOLV_LIBS $EVTLOG_NO_LIBTOOL_LIBS $SECRETSTORAGE_NO_LIBTOOL_LIBS $LD_START_STATIC -Wl,${WHOLE_ARCHIVE_OPT} $GLIB_LIBS $PCRE_LIBS $REGEX_LIBS  -Wl,${NO_WHOLE_ARCHIVE_OPT} $IVYKIS_NO_LIBTOOL_LIBS $LD_END_STATIC $LIBCAP_LIBS $DL_LIBS"
+        SYSLOGNG_DEPS_LIBS="$LIBS $BASE_LIBS $RESOLV_LIBS $EVTLOG_NO_LIBTOOL_LIBS $SECRETSTORAGE_NO_LIBTOOL_LIBS $LD_START_STATIC -Wl,${WHOLE_ARCHIVE_OPT} $GLIB_LIBS $PCRE_LIBS $REGEX_LIBS  -Wl,${NO_WHOLE_ARCHIVE_OPT} $LD_END_STATIC $IVYKIS_LIBS $LIBCAP_LIBS $DL_LIBS"
 	TOOL_DEPS_LIBS="$LIBS $BASE_LIBS $GLIB_LIBS $EVTLOG_LIBS $SECRETSTORAGE_LIBS $RESOLV_LIBS $LIBCAP_LIBS $PCRE_LIBS $REGEX_LIBS $IVYKIS_LIBS $DL_LIBS"
-	CORE_DEPS_LIBS=""
+        CORE_DEPS_LIBS="$IVYKIS_LIBS"
 
 	# bypass libtool in case we want to do mixed linking because it
 	# doesn't support -Wl,-Bstatic -Wl,-Bdynamic on a per-library basis.
diff -urNp -x '*.orig' syslog-ng-3.29.1.org/lib/logproto/tests/Makefile.am syslog-ng-3.29.1/lib/logproto/tests/Makefile.am
--- syslog-ng-3.29.1.org/lib/logproto/tests/Makefile.am	2020-08-10 20:04:39.000000000 +0200
+++ syslog-ng-3.29.1/lib/logproto/tests/Makefile.am	2021-02-28 14:27:10.230296549 +0100
@@ -23,7 +23,6 @@ lib_logproto_tests_test_findeom_CFLAGS	=
 	$(TEST_CFLAGS) \
 	-I${top_srcdir}/libtest
 lib_logproto_tests_test_findeom_LDADD	= \
-	${top_builddir}/lib/libsyslog-ng.la \
 	${top_builddir}/libtest/libsyslog-ng-test.a \
 	$(TEST_LDADD)
 lib_logproto_tests_test_findeom_SOURCES = \
